File: frontend/src/lib/request-history.ts
// Types + pure helpers for the per-request history endpoints:
//
//   GET /api/stats/v2/requests   one row per completed request in the window
//   GET /api/stats/v2/latency    TTFT / ITL / duration distributions from those rows
//
// Both are served from the request_history table the proxy feeds
// (app/stats/request_history.py). Hand-typed for the same reason as
// stats-v2.ts: the routes return `dict[str, Any]`, so openapi-typescript
// cannot reach the inner shape.
//
// Everything here is DOM-free so the chart's encoding decisions can be
// tested without rendering a chart.

import type { HistogramBuckets } from "./live-stats";
import type { StatsRange } from "./stats-v2";

// ---- wire shapes ------------------------------------------------------------

/** One completed request, as the proxy recorded it. */
export interface RequestHistoryRow {
  id: string;
  /** Epoch SECONDS when the proxy saw the request end. */
  finished_at: number;
  /** models table ROW id — what `?models=` filters on. */
  model_id: string;
  /** Served name — what a human reads. */
  model: string;
  token_name: string | null;
  client_ip: string | null;
  prompt_tokens: number;
  completion_tokens: number;
  duration_s: number;
  /** Null when no token ever arrived (abort before the first frame, or a
   *  non-streaming request, which has no first frame to time). */
  ttft_s: number | null;
  finish_reason: string | null;
  orphan: boolean;
  started_iso: string;
}

export interface HistoryCoverage {
  /** When history begins, store-wide, epoch seconds. Null when empty. */
  earliest_epoch: number | null;
  retention_days: number;
  max_rows: number;
  /** True only when history reaches back to the window start AND retention
   *  is at least as long as the window. */
  covers_window: boolean;
}

export interface RequestHistoryResponse {
  range: StatsRange;
  since_epoch: number;
  now_epoch: number;
  selected_model_ids: string[] | null;
  /** Rows in the window for the selection, before sampling. */
  total: number;
  /** 1 = every row returned; k = every k-th row, newest first. */
  stride: number;
  requests: RequestHistoryRow[];
  coverage: HistoryCoverage;
}

export interface LatencyDistribution {
  count: number;
  p50: number | null;
  p90: number | null;
  p99: number | null;
  mean: number | null;
  /** Engine-histogram wire shape (cumulative counts, `le: null` = +Inf), so
   *  the existing LatencyHistogram renderer draws it unchanged. */
  buckets: HistogramBuckets;
}

export type LatencyBasis = "window" | "last";

export interface LatencyResponse {
  basis: LatencyBasis;
  range: StatsRange;
  /** Set for basis=last. */
  n: number | null;
  since_epoch: number | null;
  selected_model_ids: string[] | null;
  count: number;
  span_s: number | null;
  oldest_epoch: number | null;
  newest_epoch: number | null;
  ttft: LatencyDistribution;
  /** MEAN inter-token latency PER REQUEST — not the engine's per-token
   *  histogram. Every label that shows it says so. */
  itl: LatencyDistribution;
  duration: LatencyDistribution;
  coverage: HistoryCoverage;
}

/** The fixed count the "last N" basis asks for. */
export const LATENCY_LAST_N = 500;

// ---- the chart's encoding ---------------------------------------------------
//
// One mark per request: time on x, duration on a log y. Colour carries ONE
// categorical field, chosen from the data (below); mark size carries generated
// tokens; finish reason is carried by the mark's SHAPE, never by colour —
// colour is already spoken for, and in retro-dark `--chat-positive` and
// `--chat-warn` are the same amber anyway. Prompt tokens and TTFT ride in
// the tooltip and the table.

/** How a request ended, folded to the three shapes the chart can draw. */
export type FinishClass = "stop" | "length" | "other";

export function finishClassOf(reason: string | null | undefined): FinishClass {
  if (reason === "stop") return "stop";
  if (reason === "length") return "length";
  return "other";
}

export const FINISH_CLASS_LABEL: Record<FinishClass, string> = {
  stop: "stop (filled)",
  length: "length (hollow)",
  other: "other / none (cross)",
};

/** Which categorical field colour encodes. */
export type ColourField = "client" | "model" | "none";

/** Ordered `text-chat-*` classes for colour groups; `currentColor` marks
 *  inherit them. Six distinct-enough tokens across both themes; the rest of
 *  the categories fold into `OTHER_CLASS`. */
export const GROUP_CLASSES: readonly string[] = [
  "text-chat-accent",
  "text-chat-negative",
  "text-chat-fg",
  "text-chat-muted",
  "text-chat-user",
  "text-chat-accent-strong",
];
export const OTHER_CLASS = "text-chat-dim";
export const OTHER_LABEL = "other";

/** The label of a client for grouping: token name first, IP as fallback. */
export function clientOf(r: Pick<RequestHistoryRow, "token_name" | "client_ip">): string {
  return r.token_name ?? r.client_ip ?? "anonymous";
}

/**
 * Decide what colour should encode, from the data.
 *
 * Colour by client is what the operator asked for, and with one dominant
 * client it earns nothing: a chart where 97% of marks share a colour reads
 * as a single series with noise. So: colour carries CLIENT when there are at
 * least two and the largest is under 95% of rows; otherwise MODEL when there
 * are at least two; otherwise nothing — one accent, and the legend says so.
 */
export function pickColourField(rows: readonly RequestHistoryRow[]): ColourField {
  if (rows.length === 0) return "none";
  const clients = countBy(rows, clientOf);
  if (clients.size >= 2 && topShare(clients, rows.length) < 0.95) return "client";
  const models = countBy(rows, (r) => r.model);
  if (models.size >= 2) return "model";
  return "none";
}

export function groupKeyOf(r: RequestHistoryRow, field: ColourField): string {
  if (field === "client") return clientOf(r);
  if (field === "model") return r.model;
  return "all";
}

export interface ColourGroup {
  key: string;
  count: number;
  /** A `text-chat-*` class. */
  className: string;
}

/**
 * The colour legend: the top groups by count, each with its class, then
 * `other` for everything past the palette. Stable across polls for the same
 * data because the order is (count desc, key asc).
 */
export function colourGroups(
  rows: readonly RequestHistoryRow[],
  field: ColourField,
): ColourGroup[] {
  if (field === "none") {
    return rows.length ? [{ key: "all", count: rows.length, className: GROUP_CLASSES[0] }] : [];
  }
  const counts = countBy(rows, (r) => groupKeyOf(r, field));
  const ordered = [...counts.entries()].sort(
    (a, b) => b[1] - a[1] || a[0].localeCompare(b[0]),
  );
  const out: ColourGroup[] = ordered
    .slice(0, GROUP_CLASSES.length)
    .map(([key, count], i) => ({ key, count, className: GROUP_CLASSES[i] }));
  const rest = ordered.slice(GROUP_CLASSES.length);
  if (rest.length > 0) {
    out.push({
      key: OTHER_LABEL,
      count: rest.reduce((s, [, c]) => s + c, 0),
      className: OTHER_CLASS,
    });
  }
  return out;
}

/** Class for one row given the legend, `OTHER_CLASS` past the palette. */
export function classFor(
  groups: readonly ColourGroup[],
  key: string,
): string {
  return groups.find((g) => g.key === key)?.className ?? OTHER_CLASS;
}

/**
 * Mark radius in px from generated tokens: sqrt so AREA tracks the count,
 * 2..6 px so thousands of marks stay legible and a giant response does not
 * eclipse its neighbours.
 */
export function markRadius(generated: number, maxGenerated: number): number {
  if (maxGenerated <= 0 || generated <= 0) return 2;
  return 2 + 4 * Math.sqrt(Math.min(1, generated / maxGenerated));
}

/** Log-scale y needs a strictly positive floor; sub-10ms durations are
 *  clamped rather than dropped so an immediate error still shows. */
export const DURATION_FLOOR_S = 0.01;

/** Nice log ticks (…0.1, 1, 10, 100…) covering [lo, hi]. */
export function logTicks(lo: number, hi: number): number[] {
  if (!(hi > 0)) return [1];
  const start = Math.floor(Math.log10(Math.max(DURATION_FLOOR_S, lo)));
  const end = Math.ceil(Math.log10(hi));
  const out: number[] = [];
  for (let e = start; e <= end; e++) out.push(10 ** e);
  return out;
}

// ---- wording ------------------------------------------------------------------

/** Seconds → "45 s", "3 h 12 m", "6 d 4 h". */
export function formatSpan(s: number | null | undefined): string {
  if (s === null || s === undefined || Number.isNaN(s) || s < 0) return "—";
  if (s < 60) return `${Math.round(s)} s`;
  const m = Math.round(s / 60);
  if (m < 60) return `${m} min`;
  const h = Math.floor(m / 60);
  const mm = m % 60;
  if (h < 24) return mm ? `${h} h ${mm} min` : `${h} h`;
  const d = Math.floor(h / 24);
  const hh = h % 24;
  return hh ? `${d} d ${hh} h` : `${d} d`;
}

/**
 * The precise statement of what a window covers, or null when it covers the
 * whole window. Wording carries the WHY: history that has not existed long
 * enough, or a retention shorter than the button — two different facts.
 */
export function coverageNote(
  coverage: HistoryCoverage | undefined,
  sinceEpoch: number | null,
  nowEpoch: number,
  human: string,
): string | null {
  if (!coverage || sinceEpoch === null) return null;
  if (coverage.covers_window) return null;
  const windowS = nowEpoch - sinceEpoch;
  if (coverage.retention_days * 86400 < windowS) {
    return `retention is ${coverage.retention_days} d, shorter than the ${human} — the window cannot be served in full`;
  }
  if (coverage.earliest_epoch === null) {
    return "no requests recorded yet";
  }
  const since = Math.max(0, nowEpoch - coverage.earliest_epoch);
  const covered = Math.min(since, windowS);
  return `history begins ${formatSpan(since)} ago — covers ${formatSpan(covered)} of the ${human}`;
}

// ---- tiny generic helpers ------------------------------------------------------

function countBy<T>(rows: readonly T[], key: (r: T) => string): Map<string, number> {
  const m = new Map<string, number>();
  for (const r of rows) {
    const k = key(r);
    m.set(k, (m.get(k) ?? 0) + 1);
  }
  return m;
}

function topShare(counts: Map<string, number>, total: number): number {
  let top = 0;
  for (const c of counts.values()) if (c > top) top = c;
  return total > 0 ? top / total : 0;
}

File: app/stats/request_history.py
"""Per-request history: the SQLite store behind the requests chart and the
latency distributions on the stats page.

WHY THIS EXISTS. Two operator complaints about the stats page had one cause:

  * "recently finished makes little sense" -- ten identical-looking rows,
    kept 15 minutes, nothing aggregated, so there was no question it answered.
  * "why are the latency charts only 5 minutes" -- the engine publishes only
    CUMULATIVE histograms, so the page kept bucket snapshots in React state and
    the "last 5 minutes" it drew was really "since you opened the tab, capped
    at 5 minutes".

Nothing persisted per-request history. This module does, and both panels read
from it. The columns are what ``app.proxy.request_registry.finished_record``
produces: the PROXY's own TTFT (first streamed frame) and duration, which exist
identically for every backend -- llama.cpp publishes no latency histogram at
all, and this is what gives GGUF models a latency panel for the first time.

WRITE PATH. ``RequestHistoryStore.record`` is called from the proxy's
``_deregister`` inside the streaming ``finally``, before the scheduler slot is
released. It therefore does NOT touch the database: it enqueues, and a
background writer drains the queue in batches. The proxy's existing guarantee
-- stats bookkeeping can never fail a request -- holds by construction here:
``record`` is synchronous, never raises, and a full queue drops the record
(counted, logged) rather than blocking a request on a busy /data volume. The
per-request accounting write (``_record_counters``) already opens its own
connection per request; this deliberately does not add a second one to the
slot-release path.

READ PATH. Module-level query functions take an open connection so the routes
can validate a selection and read on one connection, and so the queries can be
tested against a file with rows written by hand.

Everything statistical (quantiles, histograms) is pure and lives at the bottom.
"""

from __future__ import annotations

import asyncio
import logging
import math
import time
from collections.abc import Iterable, Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import aiosqlite

from app.db.database import open_db

logger = logging.getLogger(__name__)

_COLUMNS = (
    "id",
    "finished_at",
    "model_id",
    "model",
    "token_name",
    "client_ip",
    "prompt_tokens",
    "completion_tokens",
    "duration_s",
    "ttft_s",
    "finish_reason",
    "orphan",
    "started_iso",
)
_SELECT = "SELECT " + ", ".join(_COLUMNS) + " FROM request_history"
# INSERT OR IGNORE: the id is the registry's uuid, and a record that somehow
# reaches the writer twice must not turn into an IntegrityError that costs the
# whole batch.
_INSERT = (
    "INSERT OR IGNORE INTO request_history(" + ", ".join(_COLUMNS) + ") "
    "VALUES (" + ", ".join("?" for _ in _COLUMNS) + ")"
)


def _row_params(record: dict[str, Any], finished_at: float) -> tuple[Any, ...]:
    """Coerce one ``finished_record`` dict into the INSERT's parameter tuple.

    Defensive on every field: the record is built on the proxy's hot path from
    values the engine and the client controlled, and a bad value here must
    cost this one row, not the batch it travels in.
    """
    def _int(v: Any) -> int:
        try:
            return int(v or 0)
        except (TypeError, ValueError):
            return 0

    def _float(v: Any) -> float | None:
        try:
            return None if v is None else float(v)
        except (TypeError, ValueError):
            return None

    return (
        str(record.get("id") or ""),
        float(finished_at),
        str(record.get("model_id") or ""),
        str(record.get("model") or ""),
        record.get("token_name"),
        record.get("client_ip"),
        _int(record.get("prompt_tokens")),
        _int(record.get("completion_tokens")),
        _float(record.get("duration_s")) or 0.0,
        _float(record.get("ttft_s")),
        record.get("finish_reason"),
        1 if record.get("orphan") else 0,
        str(record.get("started_iso") or ""),
    )


def _row_dict(row: Iterable[Any]) -> dict[str, Any]:
    d = dict(zip(_COLUMNS, row, strict=True))
    d["orphan"] = bool(d["orphan"])
    return d


class RequestHistoryStore:
    """Queue in front of the ``request_history`` table.

    ``record`` is the only method the proxy calls. It is synchronous and
    fail-open. ``run_forever`` is the writer the app starts at boot;
    ``flush`` is the same drain, callable directly by tests and by shutdown.
    """

    def __init__(
        self,
        db_path: Path,
        *,
        queue_max: int = 10_000,
        flush_interval_s: float = 1.0,
    ) -> None:
        self.db_path = db_path
        self.flush_interval_s = float(flush_interval_s)
        self._q: asyncio.Queue[tuple[Any, ...]] = asyncio.Queue(maxsize=max(1, int(queue_max)))
        #: Records refused because the queue was full. Exposed so a test -- or
        #: an operator reading the log -- can tell "nothing happened" from
        #: "the volume was too slow to keep up".
        self.dropped = 0
        #: Rows written since start, for the same reason.
        self.written = 0
        self._last_drop_log = 0.0

    # -- write side ---------------------------------------------------------

    def record(self, record: dict[str, Any], *, finished_at: float | None = None) -> bool:
        """Enqueue one finished-request record. Never raises, never blocks.

        Returns False when the record was dropped. ``finished_at`` defaults to
        now (wall clock) -- the moment ``_deregister`` runs is the moment the
        request ended.
        """
        try:
            params = _row_params(record, time.time() if finished_at is None else finished_at)
            if not params[0]:
                return False
            self._q.put_nowait(params)
            return True
        except asyncio.QueueFull:
            self.dropped += 1
            now = time.monotonic()
            if now - self._last_drop_log > 60.0:
                self._last_drop_log = now
                logger.warning(
                    "request history: queue full, dropped %d record(s) so far; "
                    "the data volume is not keeping up",
                    self.dropped,
                )
            return False
        except Exception:  # noqa: BLE001 -- bookkeeping must never fail a request
            logger.debug("request history: could not enqueue record", exc_info=True)
            return False

    def pending(self) -> int:
        return self._q.qsize()

    def _drain(self) -> list[tuple[Any, ...]]:
        batch: list[tuple[Any, ...]] = []
        while True:
            try:
                batch.append(self._q.get_nowait())
            except asyncio.QueueEmpty:
                return batch

    async def flush(self) -> int:
        """Write everything queued right now. Returns the number of rows sent.

        A failed write logs and drops the batch. Re-queueing would turn a
        persistently broken volume into an ever-growing queue and, once full,
        into the same drop -- with the log noise multiplied.
        """
        batch = self._drain()
        if not batch:
            return 0
        try:
            async with open_db(self.db_path) as db:
                await db.executemany(_INSERT, batch)
                await db.commit()
        except Exception:
            logger.exception("request history: dropped a batch of %d record(s)", len(batch))
            return 0
        self.written += len(batch)
        return len(batch)

    async def run_forever(self) -> None:
        """Drain the queue in batches, at most once per ``flush_interval_s``.

        Waits for the first record, then sleeps one interval so a burst lands
        as one transaction rather than one fsync per request. Cancellation
        performs a final flush so a clean shutdown loses nothing queued.
        """
        try:
            while True:
                first = await self._q.get()
                await asyncio.sleep(self.flush_interval_s)
                batch = [first, *self._drain()]
                try:
                    async with open_db(self.db_path) as db:
                        await db.executemany(_INSERT, batch)
                        await db.commit()
                    self.written += len(batch)
                except asyncio.CancelledError:
                    raise
                except Exception:
                    logger.exception(
                        "request history: dropped a batch of %d record(s)", len(batch)
                    )
        except asyncio.CancelledError:
            try:
                await self.flush()
            except Exception:  # noqa: BLE001 -- shutdown must not hang on the volume
                logger.debug("request history: final flush failed", exc_info=True)
            raise


# ---------------------------------------------------------------------------
# Reads
# ---------------------------------------------------------------------------


def _model_clause(model_ids: Sequence[str] | None) -> tuple[str, tuple[str, ...]]:
    """`` AND model_id IN (...)`` for a selection, or nothing when unfiltered.

    Empty when unfiltered so the unfiltered path executes exactly the SQL it
    always would rather than a filter that happens to match everything.
    """
    if model_ids is None:
        return "", ()
    if not model_ids:
        # `IN ()` is a syntax error in SQLite and a bare 1=1 would silently
        # widen to every model -- the substitution every stats route exists to
        # prevent.
        return " AND 0", ()
    return f" AND model_id IN ({','.join('?' for _ in model_ids)})", tuple(model_ids)


@dataclass(frozen=True)
class WindowRows:
    rows: list[dict[str, Any]]
    #: How many rows the window holds for the selection, before sampling.
    total: int
    #: 1 when every row is returned; k when every k-th row (newest first) is.
    stride: int


async def query_window(
    db: aiosqlite.Connection,
    *,
    since: float,
    model_ids: Sequence[str] | None,
    limit: int,
) -> WindowRows:
    """Rows that finished at or after ``since``, newest first, at most ``limit``.

    When the window holds more than ``limit`` rows, every k-th row is returned
    rather than the newest ``limit``: this feeds a chart with a TIME axis, and
    a "newest N" cut would leave the left of the chart empty while claiming to
    show the whole window. Uniform striding keeps the span covered; the
    response says the stride so the panel can state it.
    """
    where, args = _model_clause(model_ids)
    cur = await db.execute(
        "SELECT COUNT(*) FROM request_history WHERE finished_at >= ?" + where,
        (since, *args),
    )
    total = int((await cur.fetchone() or (0,))[0] or 0)
    limit = max(1, int(limit))
    if total <= limit:
        cur = await db.execute(
            _SELECT + " WHERE finished_at >= ?" + where + " ORDER BY finished_at DESC, id DESC",
            (since, *args),
        )
        rows = await cur.fetchall()
        return WindowRows(rows=[_row_dict(r) for r in rows], total=total, stride=1)
    stride = math.ceil(total / limit)
    cur = await db.execute(
        "SELECT " + ", ".join(_COLUMNS) + " FROM ("
        "  SELECT " + ", ".join(_COLUMNS) + ", "
        "         ROW_NUMBER() OVER (ORDER BY finished_at DESC, id DESC) AS rn"
        "  FROM request_history WHERE finished_at >= ?" + where +
        ") WHERE (rn - 1) % ? = 0 ORDER BY rn ASC LIMIT ?",
        (since, *args, stride, limit),
    )
    rows = await cur.fetchall()
    return WindowRows(rows=[_row_dict(r) for r in rows], total=total, stride=stride)


async def query_window_latency(
    db: aiosqlite.Connection,
    *,
    since: float,
    model_ids: Sequence[str] | None,
) -> list[dict[str, Any]]:
    """Only the columns ``latency_summary`` needs, for EVERY row in the window.

    Deliberately uncapped: a distribution over a sample of the window is not
    the distribution of the window. Four columns a row keeps a full 7d at the
    retention cap well under a second.
    """
    where, args = _model_clause(model_ids)
    cur = await db.execute(
        "SELECT finished_at, ttft_s, duration_s, completion_tokens "
        "FROM request_history WHERE finished_at >= ?" + where,
        (since, *args),
    )
    return [
        {
            "finished_at": r[0],
            "ttft_s": r[1],
            "duration_s": r[2],
            "completion_tokens": r[3],
        }
        for r in await cur.fetchall()
    ]


async def query_last(
    db: aiosqlite.Connection,
    *,
    n: int,
    model_ids: Sequence[str] | None,
) -> list[dict[str, Any]]:
    """The newest ``n`` rows for the selection regardless of time, newest first.

    The operator's alternative basis for the latency distributions: a fixed
    number of requests does not go silent when traffic is thin, which a time
    window does.
    """
    where, args = _model_clause(model_ids)
    cur = await db.execute(
        _SELECT + " WHERE 1" + where + " ORDER BY finished_at DESC, id DESC LIMIT ?",
        (*args, max(1, int(n))),
    )
    return [_row_dict(r) for r in await cur.fetchall()]


async def earliest_finished_at(db: aiosqlite.Connection) -> float | None:
    """When history begins, across the whole store. None when empty.

    Store-wide rather than per selection on purpose: "history begins 3 h ago"
    is a statement about the store (it was added in a release, or pruned), and
    a selection with no rows is a different fact that the row count carries.
    """
    cur = await db.execute("SELECT MIN(finished_at) FROM request_history")
    row = await cur.fetchone()
    return float(row[0]) if row and row[0] is not None else None


async def prune(db: aiosqlite.Connection, *, cutoff: float, max_rows: int) -> dict[str, int]:
    """Delete rows older than ``cutoff`` and, after that, any beyond ``max_rows``.

    Age first, then count, so the count cap trims the OLDEST survivors. Does
    not commit -- the caller's transaction owns that, as with the other
    pruners.
    """
    cur = await db.execute("DELETE FROM request_history WHERE finished_at < ?", (cutoff,))
    by_age = cur.rowcount or 0
    by_count = 0
    if max_rows > 0:
        cur = await db.execute(
            "DELETE FROM request_history WHERE id IN ("
            "  SELECT id FROM request_history ORDER BY finished_at DESC, id DESC"
            "  LIMIT -1 OFFSET ?"
            ")",
            (int(max_rows),),
        )
        by_count = cur.rowcount or 0
    return {"by_age": by_age, "by_count": by_count}


# ---------------------------------------------------------------------------
# Statistics -- pure
# ---------------------------------------------------------------------------

# Bucket edges in seconds. TTFT and ITL reuse vLLM's own histogram boundaries
# so the bars line up with what vLLM's dashboards show for the same model;
# duration reuses its e2e set. The trailing +Inf bucket is encoded as ``None``
# on the wire, exactly as the engine histograms are (JSON has no infinity).
TTFT_EDGES: tuple[float, ...] = (
    0.001, 0.005, 0.01, 0.02, 0.04, 0.06, 0.08, 0.1, 0.25, 0.5, 0.75, 1.0,
    2.5, 5.0, 7.5, 10.0, 20.0, 40.0, 80.0, 160.0, 640.0, 2560.0,
)
ITL_EDGES: tuple[float, ...] = (
    0.001, 0.005, 0.01, 0.02, 0.04, 0.06, 0.08, 0.1, 0.25, 0.5, 0.75, 1.0,
    2.5, 5.0, 7.5, 10.0, 20.0, 40.0, 80.0,
)
DURATION_EDGES: tuple[float, ...] = (
    0.3, 0.5, 0.8, 1.0, 1.5, 2.0, 2.5, 5.0, 10.0, 15.0, 20.0, 30.0, 40.0,
    50.0, 60.0, 120.0, 240.0, 480.0, 960.0, 1920.0, 7680.0,
)


def quantile(sorted_values: Sequence[float], q: float) -> float | None:
    """Linear-interpolated quantile of an ascending sequence, or None if empty.

    Exact, from the samples themselves -- unlike ``hist_quantile`` over engine
    buckets, which can only interpolate inside a bucket. This is the reason to
    compute latency from per-request rows at all.
    """
    n = len(sorted_values)
    if n == 0:
        return None
    if n == 1:
        return float(sorted_values[0])
    pos = max(0.0, min(1.0, q)) * (n - 1)
    lo = int(math.floor(pos))
    hi = min(n - 1, lo + 1)
    frac = pos - lo
    return float(sorted_values[lo]) + frac * (float(sorted_values[hi]) - float(sorted_values[lo]))


def histogram(values: Iterable[float], edges: Sequence[float]) -> dict[str, Any]:
    """Cumulative buckets in the engine-histogram wire shape.

    ``le`` carries the edges plus ``None`` for +Inf; ``counts`` are cumulative
    along the boundaries, so the frontend's existing bucket renderer and
    quantile mirror consume this unchanged.
    """
    per_bucket = [0] * (len(edges) + 1)
    count = 0
    total = 0.0
    for v in values:
        count += 1
        total += v
        for i, edge in enumerate(edges):
            if v <= edge:
                per_bucket[i] += 1
                break
        else:
            per_bucket[len(edges)] += 1
    cumulative: list[int] = []
    running = 0
    for c in per_bucket:
        running += c
        cumulative.append(running)
    return {
        "le": [*edges, None],
        "counts": cumulative,
        "count": count,
        "sum": total,
    }


def distribution(values: Iterable[float], edges: Sequence[float]) -> dict[str, Any]:
    """Count, exact quantiles, mean and buckets for one latency series."""
    vals = sorted(float(v) for v in values)
    n = len(vals)
    return {
        "count": n,
        "p50": quantile(vals, 0.5),
        "p90": quantile(vals, 0.9),
        "p99": quantile(vals, 0.99),
        "mean": (sum(vals) / n) if n else None,
        "buckets": histogram(vals, edges),
    }


def itl_mean_of(row: dict[str, Any]) -> float | None:
    """Mean inter-token latency of ONE request, from the proxy's measurements.

    ``(duration - ttft) / (completion_tokens - 1)``: the decode phase divided by
    the gaps in it. This is a per-request MEAN, not the per-token distribution
    the engine's ITL histogram is -- contention inside a request averages out
    here. Every panel that shows it says "mean per request" for that reason.
    None when it cannot be computed: no first token, fewer than two tokens, or
    a decode phase of zero length (a non-streaming request has no first frame
    to time, so it lands here too).
    """
    ttft = row.get("ttft_s")
    dur = row.get("duration_s")
    n = row.get("completion_tokens") or 0
    if ttft is None or dur is None or n < 2:
        return None
    decode = float(dur) - float(ttft)
    if decode <= 0:
        return None
    return decode / (n - 1)


def latency_summary(rows: Sequence[dict[str, Any]]) -> dict[str, Any]:
    """The three distributions over a set of rows, plus their time span."""
    ttft = [float(r["ttft_s"]) for r in rows if r.get("ttft_s") is not None]
    durations = [float(r["duration_s"]) for r in rows if r.get("duration_s") is not None]
    itl = [v for v in (itl_mean_of(r) for r in rows) if v is not None]
    finished = [float(r["finished_at"]) for r in rows if r.get("finished_at") is not None]
    return {
        "count": len(rows),
        "oldest_epoch": min(finished) if finished else None,
        "newest_epoch": max(finished) if finished else None,
        "span_s": (max(finished) - min(finished)) if finished else None,
        "ttft": distribution(ttft, TTFT_EDGES),
        "itl": distribution(itl, ITL_EDGES),
        "duration": distribution(durations, DURATION_EDGES),
    }
